feat(lua): vim.fs.normalize() resolves ".", ".." (#28203)
neovim/neovim: runtime/doc/lua.txt

vim.fs.normalize({path}, {opts})                          *vim.fs.normalize()*
    Normalize a path to a standard format. A tilde (~) character at the
    beginning of the path is expanded to the user's home directory and
    environment variables are also expanded. "." and ".." components are also
    resolved, except when the path is relative and trying to resolve it would
    result in an absolute path.
    • "." as the only part in a relative path:
      • "." => "."
      • "././" => "."
    • ".." when it leads outside the current directory
      • "foo/../../bar" => "../bar"
      • "../../foo" => "../../foo"
    • ".." in the root directory returns the root directory.
      • "/../../" => "/"

    On Windows, backslash (\) characters are converted to forward slashes (/).

>     Examples: >lua

Reviewer: ```suggestion
    Example input/output: >
```

vim.fs.normalize({path}, {opts})                          *vim.fs.normalize()*
    Normalize a path to a standard format. A tilde (~) character at the
    beginning of the path is expanded to the user's home directory and
    environment variables are also expanded. `.` and `..` components are also
    resolved, except when the path is relative and trying to resolve it would
    result in an absolute path. For example: the following will not get
    resolved:
    • `.` as the only part in a relative path (e.g., `.`, `././`).

>     • Resolving `..` in a relative path leads outside the current directory

Reviewer 2: ```suggestion
    • `..` in a relative path when it leads outside the current directory
```

>     result in an absolute path. For example: the following will not get
>     resolved:
>     • `.` as the only part in a relative path (e.g., `.`, `././`).
>     • `..` in a relative path when it leads outside the current directory
>       (e.g., `../../foo`).
> 
>     NOTE: In the root directory, resolving `..` will return the root directory
>     itself.

Reviewer: - explicit input/output would be less ambiguous, less verbose, and more intuitive.
- using "" instead of backticks is more readable for this case.

```suggestion
    result in an absolute path. 
    • "." as the only part in a relative path:
        • "." => "."
        • "././" => "././"
    • ".." when it leads outside the current directory
        • "../../foo" => "../foo"
    • ".." in the root directory returns the root directory.
        • "../" => "/"
```

Author: That's way better, thanks a lot!

        vim.fs.normalize('$XDG_CONFIG_HOME/nvim/init.vim')
        -- '/Users/jdoe/.config/nvim/init.vim'

        vim.fs.normalize('~/src/nvim/api/../tui/./tui.c')
        -- '/home/jdoe/src/nvim/tui/tui.c'

        vim.fs.normalize('.')
        -- '.'

        vim.fs.normalize('./foo/bar')
        -- 'foo/bar'

        vim.fs.normalize('foo/../../../bar')
        -- '../../bar'

        vim.fs.normalize('/home/jdoe/../../../bar')

>         -- Error: Invalid path, cannot resolve `..` past root directory

Reviewer 2: I don't think this makes much sense together with "NOTE: In the root directory, resolving `..` will return the root directory itself.". `..` in root directory either errors or is resolved to the root itself.

Reviewer 3: I was gonna comment on the same thing, I got confused by this as well.

        vim.fs.normalize('$XDG_CONFIG_HOME/nvim/init.vim')
        -- '/Users/jdoe/.config/nvim/init.vim'

        vim.fs.normalize('~/src/nvim/api/../tui/./tui.c')

>         -- '/home/jdoe/src/nvim/tui/tui.c'

Reviewer: off topic, somewhat surprised that `~` is expanded without `expand_env=true` (which perhaps should be named `expand`)

Author: Fwiw, `~` is not expanded in Python's `os.path` either

Reviewer 4: Not really; that's the core part of "normalization". `expand_env` is about `$HOME` and friends (only), and it makes sense that these are independent choices.

Reviewer 3: My initial gut reaction was the same as Justin's (aka that `~` shouldn't be expanded if `expand_env` is false) although I can't with certainty say which is better/more consistent. Just wanted to share my 2 cents.

Reviewer: > `expand_env` is about `$HOME` and friends (only)

That's an artificial/accidental constraint that we can choose to change. Or we could add a separate `expand` option in the worst case.

> that's the core part of "normalization"

No, I've made this clear in multiple discussions: normalization means "the result is deterministic". That is separate from "the result is different depending on the context".

With the current behavior, the caller has to care that `~/` will be expanded even if they don't want that. For example, saving a path to a config file you probably don't want `~/` to be expanded.

        vim.fs.normalize('.')
        -- '.'

        vim.fs.normalize('./foo/bar')
        -- 'foo/bar'

        vim.fs.normalize('foo/../../../bar')
        -- '../../bar'

        vim.fs.normalize('/home/jdoe/../../../bar')
        -- /bar

        vim.fs.normalize('C:foo/../../baz')
        -- 'C:../baz'

        vim.fs.normalize('C:/foo/../../baz')

>         -- C:/baz

Reviewer: this is a very noisy way to show input/output examples.

usage examples are intended to help with the mechanics of using an interface. for input/output we can just have a table that shows input/output

```
"./foo/bar"               => "foo/bar"
"foo/../../../bar"        => "../../bar"
"/home/jdoe/../../../bar" => "/bar"
"C:foo/../../baz"         => "C:../baz"
"C:/foo/../../baz"        => "C:/baz"
```

Author: I'm not sure how to make the `=>`s align in the Vimdocs, I aligned them in the Lua docs but the vimdocs generator removes the extra spaces

Reviewer: use a codeblock

>         vim.fs.normalize('\\\\?\\UNC\\server\\share\\foo\\..\\..\\..\\bar')

Reviewer: use `[[...]]`, the escaping is just distracting from the purpose of this example.
